packages/Php70/src/Rector/FuncCall/NonVariableToVariableOnFunctionCallRector.php

```diff
@@ -0,0 +1,259 @@
+{
+    /**
+     * @var string
+     */
+    private const DEFAULT_VARIABLE_NAME = 'tmp';
+
+    /**
+     * @var Broker
+     */
+    private $broker;
+
+    /**
+     * @var ClassNaming
+     */
+    private $classNaming;
+
+    public function __construct(Broker $broker, ClassNaming $classNaming)
+    {
+        $this->broker = $broker;
+        $this->classNaming = $classNaming;
+    }
+
+    public function getDefinition(): RectorDefinition
+    {
+        return new RectorDefinition(
+            'Transform non variable like arguments to variable where a function or method expects an argument passed by reference',
+            [new CodeSample('reset(a());', '$a = a(); reset($a);')]
+        );
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [FuncCall::class, MethodCall::class, StaticCall::class];
+    }
+
+    /**
+     * @param FuncCall|MethodCall|StaticCall $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        $scope = $node->getAttribute(AttributeKey::SCOPE);
+        if (! $scope instanceof MutatingScope) {
+            return null;
+        }
+
+        $arguments = $this->getNonVariableArguments($node, $scope);
+
+        if ($arguments === []) {
+            return null;
+        }
+
+        foreach ($arguments as $key => $argument) {
+            $replacements = $this->processArgument($argument, $scope);
+
+            $current = $node->getAttribute(AttributeKey::CURRENT_STATEMENT);
+            $this->addNodeBeforeNode($replacements->assign(), $current instanceof Return_ ? $current : $node);
+            $node->args[$key]->value = $replacements->variable();
+
+            $scope = $scope->assignExpression($replacements->variable(), $scope->getType($replacements->variable()));
+        }
+
+        $node->setAttribute(AttributeKey::SCOPE, $scope);
+
+        return $node;
+    }
+
+    /**
+     * @param FuncCall|MethodCall|StaticCall $node
+     *
+     * @return array<int, Expr>
+     */
+    private function getNonVariableArguments(Node $node, Scope $scope): array
+    {
+        $arguments = [];
+
+        /** @var ParameterReflection $parameter */
+        foreach ($this->getParameters($node, $scope) as $key => $parameter) {
+            // omitted optional parameter
+            if (! isset($node->args[$key])) {
+                continue;
+            }
+
+            if ($parameter->passedByReference()->no()) {
+                continue;
+            }
+
+            $argument = $node->args[$key]->value;
+
+            if ($this->isVariableLikeNode($argument)) {
+                continue;
+            }
+
+            $arguments[$key] = $argument;
+        }
+
+        return $arguments;
+    }
+
+    /**
+     * @param FuncCall|MethodCall|StaticCall $node
+     *
+     * @return array<int, ParameterReflection>
+     */
+    private function getParameters(Node $node, Scope $scope): array
```

Comment:
As this works with reflection, it should be extracted in some service in `Reflection` namespace. Rector rules should contain as few business logic as possible, imagine it like a Controller.

Reply:
```suggestion
    private function getParameterReflections(Node $node, Scope $scope): array
```

Reply:
The service could be named `Rector\Php70\Reflection\CallReflection`